nemo/collections/asr/parts/submodules/subsampling.py

```diff
@@ -91,48 +93,40 @@ def __init__(self, subsampling, subsampling_factor, feat_in, feat_out, conv_chan
         else:
             raise ValueError(f"Not valid sub-sampling: {subsampling}!")
 
-        in_length = torch.tensor(feat_in, dtype=torch.int)
-        for i in range(self._sampling_num):
-            # length=int(in_length),
-            out_length = calc_length(
-                lengths=in_length,
-                padding=self._padding,
-                kernel_size=self._kernel_size,
-                stride=self._stride,
-                ceil_mode=self._ceil_mode,
-            )
-            in_length = out_length
-
-        out_length = int(out_length)
-        self.out = torch.nn.Linear(conv_channels * out_length, feat_out)
+        self.one = torch.tensor(1, dtype=torch.float)
+        in_length = torch.tensor(feat_in, dtype=torch.float)
+        out_length = self.calc_length(
+            in_length,
+            padding=self._padding,
+            kernel_size=self._kernel_size,
+            stride=self._stride,
+            ceil_mode=self._ceil_mode,
+        )
+        self.out = torch.nn.Linear(conv_channels * int(out_length), feat_out)
         self.conv = torch.nn.Sequential(*layers)
 
     def forward(self, x, lengths):
+        # TODO: improve the performance of length calculation
+        lengths = self.calc_length(
+            lengths,
+            padding=self._padding,
+            kernel_size=self._kernel_size,
+            stride=self._stride,
+            ceil_mode=self._ceil_mode,
+        )
         x = x.unsqueeze(1)
         x = self.conv(x)
         b, c, t, f = x.size()
-        x = self.out(x.transpose(1, 2).contiguous().view(b, t, c * f))
+        x = self.out(x.transpose(1, 2).contiguous().view(b, t, -1))
+        return x, lengths
 
-        # TODO: improve the performance of length calculation
-        new_lengths = lengths
-        # TODO: improve the performance of length calculation
+    def calc_length(self, lengths, padding, kernel_size, stride, ceil_mode):
+        """ Calculates the output length of a Tensor passed through a convolution or max pooling layer"""
+        add_pad = (padding * 2) - kernel_size
         for i in range(self._sampling_num):
-            new_lengths = calc_length(
-                lengths=new_lengths,
-                padding=self._padding,
-                kernel_size=self._kernel_size,
-                stride=self._stride,
-                ceil_mode=self._ceil_mode,
-            )
-
-        new_lengths = new_lengths.to(dtype=lengths.dtype)
-        return x, new_lengths
-
-
-def calc_length(lengths, padding, kernel_size, stride, ceil_mode):
-    """ Calculates the output length of a Tensor passed through a convolution or max pooling layer"""
-    if ceil_mode:
-        lengths = torch.ceil((lengths + (2 * padding) - (kernel_size - 1) - 1) / float(stride) + 1)
-    else:
-        lengths = torch.floor(torch.div(lengths + (2 * padding) - (kernel_size - 1) - 1, stride) + 1)
-    return lengths
+            lengths = torch.div(lengths + add_pad, stride) + self.one
```

Comment:
We need the calc_length not to be dependent on the self._sampling_num as it is going to be used as a general length calculator in our next releases. In current code, it may work but it would broke some of other experimental codes. Would it be a problem to move the loop outside of calc_length?

Reply:
It won't be a problem except for code duplication. Why moving the loop out though - we can just make it a parameter.

Reply:
I know that it is a code duplicate but the issue is that I am now working on some code changes and in those changes the parameters of each covn layer can be different from the other one like the padding. For those cases, I need to be able to pass slightly different parameters to the conv layers. For instance for downsampling of 6, we need one with striding of 3 and one with 2. Having the loop inside reduces our flexibility for future extensions.